core/src/main/java/org/apache/druid/math/expr/ApplyFunction.java

```diff
@@ -114,6 +150,14 @@ default boolean hasArrayOutput(LambdaExpr lambdaExpr)
   @Nullable
   ExpressionType getOutputType(Expr.InputBindingInspector inspector, LambdaExpr expr, List<Expr> args);
 
+  class ValidationException extends IllegalArgumentException
```

Comment:
It's nice to have a specific exception for validation!
Do you think it would be better to have a single `ExpressionValidationException` that is used across Druid (or across all Druid expressions), rather than having separate `Function.ValidationException` and `ApplyFunction.ValidationException`?

Reply:
reworked stuff to share exception handling between `Function`, `ApplyFunction`, and `ExprMacro`, and consolidated the exceptions to `ExpressionValidationException` and `ExpressionProcessingException`.